packages/components/docs/immer.md

```diff
@@ -0,0 +1,235 @@
+
+```ts
+import { produce } from 'immer';
+
+let circle = produce(new Circle(5), () => {});
+circle.radius = 10; // circle.radius is now 10! I thought using produce made it immutable!
+```
+
+To make a class immutable with Immer you first annotate the class with a [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)
+provided by Immer called `immerable`:
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
+```
+
+What does this symbol do? To quote the Immer docs:
+
+> Every other object must use the `immerable` symbol to mark itself as compatible with Immer.
+When one of these objects is mutated within a producer, its prototype is preserved between copies.
+
+Now this class is ready to be used with Immer. Let's try again using `produce`:
+
+```ts
+import { produce } from 'immer';
+
+// without using produce the instance is still mutable
+let circle = new Circle(5);
+circle.radius = 20; // radius now 20.
+
+// with produce the instance is now immutable
+circle = produce(new Circle(5), () => {});
+circle.radius = 10; // non-strict mode: fails silently. Radius still 5.
+circle.radius = 10; // strict mode: Run-time error: Cannot assign to read only property 'radius' of object '#<Circle>'
+```
+
+Instances of this class declared via `produce` can only be mutated via `produce`. When an instance is passed through
+Immer's `produce` function it will [freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze)
+the object (when `immer.setAutoFreeze(true)`). Depending on the strict mode, any attempts to explicitly modify the object
+will either fail to modify or throw a run-time error.
+
+Now, let's actually make an update to the immutable instance with Immer:
+
+```ts
+import { produce } from 'immer';
+
+// Create the initial instance
+const circle = produce(new Circle(5), () => {});
+
+// Mutate and copy from `produce`
+const newCircle = produce(circle, (draft) => {
+    draft.radius = 10;
+});
+console.log(circle.radius);    // 5
+console.log(newCircle.radius); // 10
+```
+
+The instance is now immutable, a mutated copy can be made via `produce`, and we have run-time safety via `Object.freeze`.
+
+#### Compile-time safety
+
+The main advantage of using TypeScript is that it compiles your code to ensure correctness before you run it. Additionally,
+the run-time safety is great but it still may be something that only occurs in production if say our test coverage misses
+a code path.
+
+To get started, let's first declare all the properties on the `Circle` class as [`readonly`](https://www.typescriptlang.org/docs/handbook/classes.html#readonly-modifier).
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    readonly radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
```

Comment:
Also it might be worth doing an example with TypeScript Readonly<T> as I think this would be more useful than individually labeling each property as readonly.

https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlyt

Reply:
Oh nevermind, it's only useful for instance creation, not definition.  Thought maybe it could be used in the constructor but I guess not.  Not that helpful